### ISSUE TITLE
Mongodb source should only send new added data to the sink but not all the documents

### ISSUE THREAD

**Issue by AskaJason** [spring-cloud-stream-app-starters/mongodb#3]
**using maven://org.springframework.cloud.stream.app:mongodb-source-rabbit:1.2.1.BUILD-SNAPSHOT from repo: http://repo.spring.io/libs-snapshot-local**

By using this DSL:
`stream create --name mongodb7hdfs7 --definition "mongodb --database=sourcedb --host=172.20.74.91 --fixed-delay=5 --cron='*/10 * * * * *' --initial-delay=5 --max-messages=1 --collection=sourceCol | log" –deploy`

The mongodb source will send all the documents of the collection to the sink but not only the new added document. So in the monitored log console, every 10 seconds all the documents will be printed but we expect to see only the new added document should be printed.

**Comment by AskaJason** [comment]
@sobychacko do you know if it's an issue or just incorrect configuration?

**Comment by artembilan** [comment]
I suggest to read Spring Integration [Reference Manual](http://docs.spring.io/spring-integration/reference/html/mongodb.html#mongodb-inbound-channel-adapter) on the matter carefully and pay attention to that `TransactionSynchronizationFactory` stuff.

The idea is to customize this app starter with the particular custom logic to avoid duplicates.

You may consider to fetch only documents which not older them some date though....


Although I think there should be some level of improvement like `updateQuery` to be used from the `TransactionSynchronizationFactory` for the logic you might expect already...

**Comment by AskaJason** [comment]
Hi @artembilan , thank you for the suggestion. 
But currently the behavior of the mongodb source is that it is continuously polling the entire documents and send to the sink again and again once the stream is deployed. And this will never stop unless it is undployed.
Or is there any way to configure the mongodb source only polling once after it is deployed? At least batch mongodb source ingestion is better than ingest entire documents to sink in endless loop.

Can you provide a batch or streaming ingestion sample based on mongodb source without any code change? Otherwise the mongodb source really cannot be used in any actual case.

**Comment by artembilan** [comment]
No, I agree, there is no such a scenario (at least I don't see it) where the current out-of-the-box  MongoDB Source may work as you expect. You definitely need to change the code of the existing app and provide an appropriate logic.

Here is Doc how to proceed: http://docs.spring.io/spring-cloud-stream-app-starters/docs/Bacon.RELEASE/reference/html/_introduction.html#_creating_custom_artifacts.

I won't mind (and even be glad) if someone else will share his own wisdom or at least experience how to use this MongoDB Source...

**Comment by AskaJason** [comment]
Hi @artembilan , thank you! I will try to figure out how to do the code change.
Again, much appreciate your helpful information!

**Comment by artembilan** [comment]
Reopen, since it is really something we have to reconsider for the out-of-the-box app and eliminate its uselessness.

**Comment by AskaJason** [comment]
Hi @artembilan , in my case, the expected use case for the mongodb source app could be like:
1. first time poll all the documents returned by the query
2. persist the value of the field (ie: updated_timestamp) defined by the user as the '**offset**' in H2 or MySQL
3. when the source mongodb is triggered by the cron, poll all the documents after the offset then suspend and wait for next time cron trigger
4. loop step 2 and 3 until the stream is undeployed
So the source source app will poll documents increasedly and regularly for example every 2 seconds. This can be used in steaming ingestion or real time processing such as IOT case. As mongodb has good performance to return queried document from massive data, it is always used to store the sensor info of IOT device for the source of the real time ingestion or ETL.

Plus: for the 'offset', user could also provide it by himself for initial running. If it's empty, then the first time polling will return all the documents by the defined mongodb query This 'offset' mechanism might also be the common expected feature for all the steam source apps. 

Spring CDF is a very cool thing with cloud native. Definitely it would bring value to big data market and users. Many thanks for your contributions!

**Comment by artembilan** [comment]
OK, @AskaJason ,

What you describe is exactly custom logic. That for you it feels like a good fit to handle some offset and more over via RDBMS. We definitely can't make it as an out-of-the-box feature. It doesn't cover all the possible use-case.

The best choice what I suggest is something like `updateQuery`. Fully similar what we have in the JDBC Source: https://github.com/spring-cloud-stream-app-starters/jdbc/blob/master/spring-cloud-starter-stream-source-jdbc/src/main/java/org/springframework/cloud/stream/app/jdbc/source/JdbcSourceProperties.java#L38

I won't mind for any other opinion, that is just me not feeling as a good choice to bring RDMBS into the MongoDB microservice...

**Comment by egfesser** [comment]
It's been a few months since the last post on this thread. @AskaJason - I'm wondering if you ever made progress on the code change you mentioned. @artembilan - I just want to make sure I understood your earlier comments. It sounded like you were looking to reconsider this functionality for the out-of-the-box app, but you later commented that it can't be implemented for the out-of-the-box app. Can you please update your current thoughts? I'm also wondering what purpose is served when the same data is repeatedly re-streamed (in my case via rabbitmq). In other words, why would a jdbc source be treated differently than a mongodb source? - The jdbc source I set up seems to work as expected, as does the mongodb sink and jdbc sink, just not the mongodb source.

**Comment by gogi2811** [comment]
@artembilan I am also having similar issue with MongoDb source.
I am trying to get the data for last 24hrs form mongodb.
Neither the query nor SPEL expression is working.
PFB my query and SPEL expression I am using.
Would request you share your thoughts on how to get desired data from mongoDB source.I am usong PCF as my env.
 **query**

```
mongodbv13 --uri=mongodb://user:pass@x.xx.xx.xx:27017/test --fixed-delay=5 --collection=test --query="{date:{$gt: new Date((new Date().getTime() - (24 * 60 * 60 * 1000)))}}" | log
```
**SPEL query I am trying**
```
mongodbv13 --uri=mongodb://user:pass@x.xx.xx.xx:27017/test --fixed-delay=5 --collection=test --query-expression=@Query('date':{'$gte':new Date((new Date().getTime() - (24 * 60 * 60 * 1000)))}}) | log
```

Error I am getting with SPEL query
 

```
 Property: mongodb.queryExpression
 2018-05-20T11:01:11.99+0530 [APP/PROC/WEB/0] OUT     Value: @Query(value="{'date':{'$gte':new Date((new Date().getTime() - (24 * 60 * 60 * 1000)))}}")
2018-05-20T11:01:11.99+0530 [APP/PROC/WEB/0] OUT     Reason: Failed to convert property value of type 'java.lang.String' to required type 'org.springframework.expression.Expression' for property 'queryExpression'; nested exception is org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [java.lang.String] to type [org.springframework.expression.Expression]
```

**Comment by gogi2811** [comment]
@artembilan  I was able to resolve the issue . @AskaJason this is what you might have also done .
 
I created a Stored function in mongo db which returns the iso data for last day.
**Mongo Db Stored Function**
`db.system.js.save({
    _id: "getLastDayDocs",
    value: function () {
        return new Date((new Date().getTime() - (24 * 60 * 60 * 1000)));
    }
})`

I used the below Query in MongoDB source as below 
`mongodb --uri=mongodb://xxxx:xxxx@xxxxxx:27017/xxxxx --fixed-delay=5 --initial-delay=5 --collection=xxx **--query='{$where: "function() { return (this.date > getLastDayDocs())}"}'** | xxxxx: Processor | xxxxxxxx: mongodb --uri=mongodb://xxxxx:xxx@xxxxxx:27017/xxxxx --collection=xxx`

And the problem was solved.
Hope this helps. :)

**Comment by artembilan** [comment]
@gogi2811 ,

I'm glad that you have found a solution, but what I can say about your queries and your solution that it isn't related to what we decided to do in this ticket. I would even say that your problem is more like StackOverflow question.

My point here that we have to add an `update` option to the `MongoDbMessageSource`, which we don't have there so far. That is already Spring Integration concern. And one more argument for that approach that we have a similar `JdbcPollingChannelAdapter` which does have an `update` option to be performed after selecting.

Everything else you would like to discuss is just an unnecessary noise in the ticket, although I admit that it is useful (at least for me) to how people try to workaround a limitation.

I think your ` --query` sample is somehow wrong by the MongoDB query syntax, or there is not enough quotes for SCDF to parse the value properly...

The sample with the `--query-expression` is fully unclear to me. I don't know what is `@Query` and that is really a surprise that you have a `ConverterNotFoundException`. You might use some old version of the MongoDB source.

Anyway all of that has to go to StackOverflow and here we have to concentrate on the `update` property. 

Just IMO and I would be glad to hear other opinions: I don't feel the subject properly to be sure what and how should work...

**Comment by iceddante** [comment]
@artembilan artembilan, as per our conversation on SO I will try to make some progress on this.

**Comment by artembilan** [comment]
Thanks, @iceddante .
Although, please, consider to apply an `update` fix into the [Spring Integration](https://github.com/spring-projects/spring-integration) project, not here.

Let's start from [JIRA](https://jira.spring.io/projects/INT) though to track all the resources and discussions!

P.S. An SO you have mentioned: https://stackoverflow.com/questions/51676101/implementing-a-mongodb-inbound-flow-with-spring-integration
